### PR TITLE
allow empty auth definitions in parser

### PR DESCRIPTION
**Pull request by spolti**
fixes https://github.com/serverlessworkflow/sdk-go/issues/110

Signed-off-by: spolti <filippespolti@gmail.com>

**Many thanks for submitting your Pull Request :heart:!**

**What this PR does / why we need it**:

**Special notes for reviewers**:

**Additional information (if needed):**

### CODE REVIEW

File: model/auth.go [hunk 1]
```diff
@@ -105,9 +106,12 @@ func (a *AuthDefinitions) UnmarshalJSON(b []byte) error {
 		return a.unmarshalFile(b)
 	case '[':
 		return a.unmarshalMany(b)
+	case '{': // single values are not supported, should we support it?
```
Comment (lsytj0413) [inline model/auth.go:109]: Can we add an json/yaml file in testdata to validate this?

File: model/auth.go [hunk 2]
```diff
@@ -105,9 +106,12 @@ func (a *AuthDefinitions) UnmarshalJSON(b []byte) error {
...
+		return fmt.Errorf("authDefinitions value '%s' is not supported, it must be an object or string", string(b))
+	default:
```
Comment (lsytj0413) [inline model/auth.go:111]: If the user configured it as:
`
"auth": 123
`

There will not have error happened?

**Comment by lsytj0413** [comment]
The problem is not cause by `AuthDefinitions.Unmarshal` implement, according to the specificaiton, `auth` field is an array of object, which like:
image


But in the source code, we place an `Defs` field under Auth Definition struct. So after we marshal object to json, it looks like:
```
{
	"id": "applicantrequest",
	"name": "Applicant Request Decision Workflow",
	"description": "Determine if applicant request is valid",
	"version": "1.0",
	"start": {
		"stateName": "CheckApplication"
	},
	"specVersion": "0.7",
	"expressionLang": "jq",
	"auth": {
		"Defs": [{
			"name": "testAuth",
			"scheme": "bearer",
			"properties": {
				"token": "test_token"
			}
		}, {
			"name": "testAuth2",
			"scheme": "basic",
			"properties": {
				"username": "test_user",
				"password": "test_pwd"
			}
		}]
	},
	"states": [{
		"name": "CheckApplication",
		"type": "switch",
		"defaultCondition": {},
		"dataConditions": [{
			"condition": "${ .applicants | .age \u003e= 18 }",
			"transition": {
				"nextState": "StartApplication"
			}
		}, {
			"condition": "${ .applicants | .age \u003c 18 }",
			"transition": {
				"nextState": "RejectApplication"
			}
		}],
		"timeouts": {}
	}, {
		"name": "StartApplication",
		"type": "operation",
		"end": {
			"terminate": true
		},
		"actions": [{
			"subFlowRef": {
				"workflowId": "startApplicationWorkflowId",
				"invoke": "sync",
				"onParentComplete": "terminate"
			},
			"sleep": {},
			"actionDataFilter": {
				"useResults": true
			}
		}]
	}, {
		"name": "RejectApplication",
		"type": "operation",
		"end": {
			"terminate": true
		},
		"actionMode": "sequential",
		"actions": [{
			"functionRef": {
				"refName": "sendRejectionEmailFunction",
				"invoke": "sync"
			},
			"sleep": {},
			"actionDataFilter": {
				"useResults": true
			}
		}]
	}],
	"functions": [{
		"name": "sendRejectionEmailFunction",
		"operation": "http://myapis.org/applicationapi.json#emailRejection"
	}],
	"retries": [{
		"name": "TimeoutRetryStrategy",
		"delay": "PT1M",
		"maxAttempts": "5",
		"jitter": 0
	}]
}
```

It's not equivalent. I think we should fix the struct field layout.

**Comment by spolti** [comment]
needed vor v08 https://github.com/serverlessworkflow/sdk-go/issues/70

File: model/auth.go [hunk 3]
```diff
@@ -99,15 +100,35 @@ func (a *AuthDefinitions) UnmarshalJSON(b []byte) error {
...
+		case b[index] == '"':
+			return a.unmarshalFile(b)
+		case b[index] == '[':
+			return a.unmarshalMany(b)
+		case b[index] == '{': // single values are not supported, should we support it?
+			return fmt.Errorf("authDefinitions value '%s' is not supported, it must be an object or string", string(b))
+		}
+	}
+	return nil
+	// See if we can guess based on the first character
```
Comment (spolti) [inline model/auth.go:119]: I'll remove that later

File: model/auth.go (hunk 2 above)
Comment (spolti) [reply model/auth.go]: Tried a different approach, please take a look.

**Comment by spolti** [comment]
About the defs problem, there is another problem as well, see:

https://gist.github.com/spolti/ac43ce2cc5a51532987259db4f9fa5b7
Will address it on another issue: https://github.com/serverlessworkflow/sdk-go/issues/126

File: model/auth.go [hunk 4]
```diff
@@ -99,15 +100,35 @@ func (a *AuthDefinitions) UnmarshalJSON(b []byte) error {
...
-	case '"':
-		return a.unmarshalFile(b)
-	case '[':
-		return a.unmarshalMany(b)
-	}
+	for index, char := range b {
+
+		switch {
+		case b[index] >= '0' && b[index] <= '9':
+			fmt.Print(string(char))
```
Comment (lsytj0413) [inline model/auth.go:107]: We shoule delete this.

Comment (spolti) [reply model/auth.go:107]: maybe, this makes sure that numbers can't be parsed, as you've asked before :)

File: model/workflow.go [hunk 5]
```diff
@@ -108,7 +108,7 @@ type BaseWorkflow struct {
 	// Auth definitions can be used to define authentication information that should be applied to resources defined in the operation
 	// property of function definitions. It is not used as authentication information for the function invocation,
 	// but just to access the resource containing the function invocation information.
-	Auth AuthDefinitions `json:"auth,omitempty"`
+	Auth *AuthDefinitions `json:"auth,omitempty"`
```
[the remark this answers is not in the document]
Comment (spolti) [reply model/workflow.go:111]: yes, tried already, the problem is, if we do that we can't have strings :)
maybe declaring it as interface might help, let's see..

[the remark this answers is not in the document]
Comment (lsytj0413) [reply model/workflow.go:111]: I think we can do it in this way.
`
// First, we redefine the []Auth
type AuthArray []Auth

// Second, change struct field
Auth AuthArray

// Third, define UnmarshalJSON for AuthArray, and implement it
func (a *AuthArray) UnmarshalJSON(data []byte) error {
// Unmarshal it as string or array
}
`

[the remark this answers is not in the document]
Comment (spolti) [reply model/workflow.go:111]: I wasn't able to get it working with this approach, however with interface it is fully working, including the fix for this issue and https://github.com/serverlessworkflow/sdk-go/issues/126

[the remark this answers is not in the document]
Comment (spolti) [reply model/workflow.go]: @davidesalerno  f.y.i

File: model/workflow.go [hunk 6]
```diff
@@ -108,7 +108,7 @@ type BaseWorkflow struct {
 	// Auth definitions can be used to define authentication information that should be applied to resources defined in the operation
 	// property of function definitions. It is not used as authentication information for the function invocation,
 	// but just to access the resource containing the function invocation information.
-	Auth AuthDefinitions `json:"auth,omitempty"`
+	Auth interface{} `json:"auth,omitempty"`
```
Comment (lsytj0413) [inline model/workflow.go:111]: interface is confusing to the user, i have give an implement example in https://github.com/spolti/sdk-go/pull/1，PTAL

Comment (spolti) [reply model/workflow.go]: Nice :)
I think it is ready to go now.

File: model/workflow.go [hunk 7]
```diff
@@ -130,6 +184,34 @@ func (w *Workflow) UnmarshalJSON(data []byte) error {
 	if err := json.Unmarshal(data, &workflowMap); err != nil {
 		return err
 	}
+
+	// if len(workflowMap["auth"]) > 0 {
```
Comment (lsytj0413) [inline model/workflow.go:188]: should we just delete this？

Comment (spolti) [reply model/workflow.go:188]: oops, we should, I've forgot to delete :)

**Review by ricardozanini — APPROVED** [review]

**Pull request merged**